File: src/main/java/io/github/jsnimda/common/config/options/ConfigEnum.java
package io.github.jsnimda.common.config.options;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

import io.github.jsnimda.common.config.ConfigOptionBase;
import io.github.jsnimda.common.config.IConfigOptionPrimitive;
import io.github.jsnimda.common.config.IConfigOptionToggleable;

public class ConfigEnum<E extends Enum<E>> extends ConfigOptionBase implements IConfigOptionPrimitive<E>, IConfigOptionToggleable {

  private final E defaultValue;
  private E value;

  public ConfigEnum(E defaultValue) {
    this.defaultValue = defaultValue;
  }

  @Override
  public void toggleNext() {
    E[] values = value.getDeclaringClass().getEnumConstants();
    value = values[(value.ordinal() + 1) % values.length];
  }

  @Override
  public void togglePrevious() {
    E[] values = value.getDeclaringClass().getEnumConstants();
    value = values[(value.ordinal() + values.length - 1) % values.length];
  }

  @Override
  public E getValue() {
    return value;
  }

  @Override
  public E getDefaultValue() {
    return defaultValue;
  }

  @Override
  public void setValue(E value) {
    this.value = value;
  }

  @Override
  public JsonElement toJsonElement() {
    return new JsonPrimitive(value.name());
  }

  @Override
  public void fromJsonElement(JsonElement element) {
    if (element.isJsonPrimitive() && element.getAsJsonPrimitive().isString()) {
      value = Enum.valueOf(value.getDeclaringClass(), element.getAsString());
    } else {
      // TODO fail log
    }
  }

}

File: src/main/java/io/github/jsnimda/common/config/options/ConfigEnum.kt
package io.github.jsnimda.common.config.options

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import io.github.jsnimda.common.Log
import io.github.jsnimda.common.config.ConfigOptionBase
import io.github.jsnimda.common.config.IConfigOptionPrimitive
import io.github.jsnimda.common.config.IConfigOptionToggleable
import io.github.jsnimda.common.next
import io.github.jsnimda.common.previous

class ConfigEnum<E : Enum<E>>(override val defaultValue: E) : ConfigOptionBase(), IConfigOptionPrimitive<E>, IConfigOptionToggleable {
  override var value = defaultValue

  override fun toggleNext() {
    value = value.next()
  }

  override fun togglePrevious() {
    value = value.previous()
  }

  override fun toJsonElement(): JsonElement {
    return JsonPrimitive(value.name)
  }

  override fun fromJsonElement(element: JsonElement) {
    resetToDefault()
    try {
      value = java.lang.Enum.valueOf(value.declaringClass, element.asString)
    } catch (e: Exception) {
      Log.warn("[invprofiles.common] Failed to set config value for '$key' from the JSON element '$element'")
    }
  }

}